addon/PlacesProvider.js

```diff
@@ -667,46 +667,78 @@ Links.prototype = {
+   * Get all links visited in the past 4 days or default to last 20 links.
    * Last 20 links is used as fallback in case there is no recent history.
    *
    * @param {Object} options
-   *        {Integer} limit: Maximum number of results to return. Max 20
    *
    * @returns {Promise} Returns a promise with the array of links as payload.
    */
   getRecentlyVisited: Task.async(function*(options = {}) {
     let {ignoreBlocked} = options;
 
     let blockedURLs = ignoreBlocked ? [] : this.blockedURLs.items().map(item => `"${item}"`);
+    let params = {limit: LINKS_QUERY_LIMIT};
 
-    let sqlQuery = `SELECT p.url as url,
-                           p.guid as guid,
-                           f.data as favicon,
-                           f.mime_type as mimeType,
-                           p.title as title,
-                           p.frecency as frecency,
-                           p.visit_count as visitCount,
-                           p.last_visit_date / 1000 as lastVisitDate,
-                           "history" as type,
-                           b.id as bookmarkId,
-                           b.guid as bookmarkGuid,
-                           b.title as bookmarkTitle,
-                           b.lastModified / 1000 as lastModified,
-                           b.dateAdded / 1000 as bookmarkDateCreated
-                    FROM moz_places p
-                    LEFT JOIN moz_bookmarks b
-                      ON b.fk = p.id
-                    LEFT JOIN moz_favicons f
-                      ON p.favicon_id = f.id
-                    WHERE datetime(p.last_visit_date / 1000 / 1000, 'unixepoch') > datetime('now', '${HIGHLIGHTS_THRESHOLDS.ageLimit}')
-                    AND p.url NOT IN (${blockedURLs})
-                    AND p.title NOT NULL
-                    AND p.rev_host NOT IN (${REV_HOST_BLACKLIST})`;
+    let sqlQuery = `SELECT DISTINCT * FROM (
+                      SELECT * FROM (
+                        SELECT p.url as url,
+                          p.guid as guid,
+                          f.data as favicon,
+                          f.mime_type as mimeType,
+                          p.title as title,
+                          p.frecency as frecency,
+                          p.visit_count as visitCount,
+                          p.last_visit_date / 1000 as lastVisitDate,
+                          "history" as type,
+                          b.id as bookmarkId,
+                          b.guid as bookmarkGuid,
+                          b.title as bookmarkTitle,
+                          b.lastModified / 1000 as lastModified,
+                          b.dateAdded / 1000 as bookmarkDateCreated
+                        FROM moz_places p
+                        LEFT JOIN moz_bookmarks b
+                          ON b.fk = p.id
+                        LEFT JOIN moz_favicons f
+                          ON p.favicon_id = f.id
+                        AND p.url NOT IN (${blockedURLs})
```

Comment:
This `AND` used to be part of the `WHERE` but happens to work in the `ON`. Not sure if it would make sense to have a shared query template accepting an optional extra `WHERE` and `LIMIT` to avoid these slight differences.